exonum/src/runtime/mod.rs

```diff
@@ -227,85 +228,93 @@ impl From<RuntimeIdentifier> for u32 {
-    /// Indeed, committing the service will not follow if the alternative block proposal without
-    /// the service instantiation was accepted. If the `commit_service` call *is* performed, it is
-    /// guaranteed to be performed in the closest committed block, i.e., before the nearest
-    /// `Runtime::after_commit()`.
-    ///
-    /// The dispatcher does not route transactions and `before_transactions` / `after_transactions`
-    /// events to the service until after `commit_service()` is called with the same instance spec.
-    ///
-    /// The runtime should discard the instantiated service instance after completing this method,
-    /// unless there are compelling reasons to retain it (e.g., creating an instance takes very
-    /// long time).
-    /// Alternatively, "garbage" services may be removed from `Runtime` in `after_commit`
+    /// The constructor runs *exactly once* during the blockchain lifetime for each successfully
+    /// initialized service instance. That is to say, the constructor is *not* called on a node restart.
+    ///
+    /// At the same time, when `start_adding_service` is called,
+    /// there is no guarantee that the service will eventually get to the blockchain via `commit_service`.
+    /// If the consensus accepts an alternative block proposal without the service inside,
+    /// commit of the service will fail.
+
+    /// The `commit_service` call always takes place
+    /// in the closest committed block, i.e., before the nearest `Runtime::after_commit()`.
+    /// The dispatcher routes transactions and `before_transactions` / `after_transactions`
+    /// events to the service only after `commit_service()` is called with the same instance specification.
+    ///
+    /// The runtime should discard the instantiated service instance after completing this method.
+    /// Otherwise, if the service is successfully committed in the block, it will duplicate the one instantiated
+    /// in the runtime. There may be compelling reasons for the runtime to retain the instantiated
+    /// service. For example, if creating an instance takes very long time.
+    /// In this case, the "garbage" services may be removed from the runtime in `after_commit`
     /// because of the time dependence between `commit_service` and `after_commit` described above.
-    /// The runtime should commit long-term resources for the service after a `commit_service()` call.
-    /// Since discarded instances persist their state in a discarded fork, no further action
-    /// is required to remove this state.
     ///
-    /// # Return value
+    /// According to the principle of the consensus algorithm, the process of forming a new block
+    /// takes place in RAM. A patch with changes is then applied to the database. If a block proposal
+    /// is not accepted, a fork with the patch containing this proposal is discarded from RAM.
+    /// The service instances, already discarded by the runtime, persist their state only in the
```

Comment:
It does not matter whether the service instance is discarded by the runtime - the service state is placed within the patch anyway (this is not something a runtime can influence). Thus, I'd remove "already discarded by the runtime". I'd also remove the preceding 3 sentences because they are too generic / don't really concern the documented method.